src/weighted_map.h

```diff
@@ -0,0 +1,193 @@
+    {
+        while (fromIndex > 0) {
+            fromIndex = parent(fromIndex);
+            nodes[fromIndex].sumOfDescendantWeights += weightDelta;
+        }
+    }
+
+    // For a given random weight, this method recursively finds the index of the
+    // correct entry. This is used by WeightedMap::takeRandom().
+    size_t findByWeight(size_t fromIndex, W weightToFind) const
+    {
+        W leftWeight = getWeightAt(leftChild(fromIndex));
+        W rightWeight = getWeightAt(fromIndex) - getWeightAt(rightChild(fromIndex));
+        // On Left
+        if (weightToFind < leftWeight) {
+            return findByWeight(leftChild(fromIndex), weightToFind);
+        }
+        // Found
+        if (weightToFind < rightWeight) {
+            return fromIndex;
+        }
+        // On Right
+        return findByWeight(rightChild(fromIndex), weightToFind - rightWeight);
+    }
+
+public:
+    WeightedMap() {}
+
+    // Return the total weight of all entries in the map.
+    W getTotalWeight() const
+    {
+        return getWeightAt(0);
+    }
+
+    // Return true when the map has no entries.
+    bool empty() const
+    {
+        return nodes.empty();
+    }
+
+    // Return the number of entries.
+    size_t size() const
+    {
+        return nodes.size();
+    }
+
+    // Return false if the key already exists in the map.
+    // Otherwise, add an entry mapping `key` to `value` with the given weight,
+    // and return true. The weight must be positive.
+    bool add(K key, V value, W weight)
+    {
+        assert(W() < weight);
+        if (indexMap.count(key) > 0) {
+            return false;
+        }
+        size_t index = nodes.size();
+        nodes.push_back(Node {
+            .key = key,
+            .value = value,
+            .weight = weight,
+            .sumOfDescendantWeights = W(),
+        });
+        indexMap[key] = index;
+        backPropagate(index, weight);
+        return true;
+    }
+
+    // If the given key is not present in the map, return std::nullopt.
+    // Otherwise, remove that key's entry and return its associated value.
+    std::optional<V> remove(K key)
+    {
+        if (indexMap.count(key) == 0) {
+            return std::nullopt;
+        }
+
+        size_t removeIndex = indexMap.at(key);
+        V removeValue = nodes.at(removeIndex).value;
+
+        size_t lastIndex = nodes.size()-1;
+        Node lastNode = nodes.at(lastIndex);
+        W weightDelta = lastNode.weight - nodes.at(removeIndex).weight;
+        backPropagate(lastIndex, -lastNode.weight);
+
+        if (removeIndex < lastIndex) {
+            nodes[removeIndex].key = lastNode.key;
+            nodes[removeIndex].value = lastNode.value;
+            nodes[removeIndex].weight = lastNode.weight;
+            // nodes[removeIndex].sumOfDescendantWeights should not change here.
+            indexMap[lastNode.key] = removeIndex;
+            backPropagate(removeIndex, weightDelta);
+        }
+
+        indexMap.erase(key);
```

Comment:
Yes this is okay because the insertion `indexMap[lastNode.key] = removeIndex;` [does not invalidate the iterator](https://en.cppreference.com/w/cpp/container#Iterator_invalidation).